### ISSUE TITLE
backend: pcloud: support block level sync (upload only)

### ISSUE THREAD

**Issue by gfelbing** [rclone/rclone#7896]
#### The associated forum post URL from `https://forum.rclone.org`

- Block level sync in general was requests in the forum [here](https://forum.rclone.org/t/pcloud-block-level-sync/43455)
- It helps fixing the the rate limit issues discussed [here](https://forum.rclone.org/t/how-to-handle-pcloud-rate-limit/31279/19):
  - The described work around only helps when uploading multiple smaller files by increasing `--transfers`
  - For single large files, I haven't found any workaround.
  - Block level sync seems to be slightly faster, but the main point is: With this, we can resume failed uploads of large files (e.g. due to timeouts caused by rate limit)

#### What is your current rclone version (output from `rclone version`)?

- master

#### What problem are you are trying to solve?

- Implement object.SetModTime to reduce unecessary re-uploads on mtime mismatch
  - This is an undocumented API feature (reverse-engineered), using server-side copy with same target, only changing metadata.
  - This is necessary for block level support, otherwise uploaded files will have mtime from time of upload.
- Implement new option `upload_chunk_size`
  - Enables chunk-wise uploads of files.
  - Successful uploaded chunks are skipped when upload needs retry, thanks to block-wise hashing capability.

Focusing on the upload in this issue, block level download can easily be added in follow-up.

---

tl;dr

Uploading large files to pcloud using the uploadfile API is error prone, as any connection reset / package loss will cause the full file to be retried. 
Additionally, I can see that pcloud applies some sort of rate limiting, causing large uploads to be throttled down. 
(Their "support" claims that there is no such rate limit and blames my ISP. But by increasing the `--transfers` with multiple files, I can multiply the throughput by amount of transfers until I reach my ISPs upload lim. Also 32kbps is such a round number, I doubt that its coincidence.)

In my case, this renders it almost impossible to upload any file larger than ~200MB via rclone.

Fortunately, pcloud offers a [fileops API](https://docs.pcloud.com/methods/fileops/), which enables chunked uploads on block level 🎉
PR #7895  adds an advanced option `upload_chunk_size` to the pcloud backend.
When set, files will be uploaded in chunks using the fileops API. Otherwise, current behavior is used (backwards compatibility).

#### How do you think rclone should be changed to solve that?

I created a draft PR #7895 that implements it. 
Feature is drafty, but works on my machine :). Before I invest more time in writing tests and documentation, I'd appreciate a first review, whether the feature itself / the way I built it goes into the right direction.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Great idea :-)

I gave your PR an initial review.

**Comment by gfelbing** [comment]
With the current version, I was finally completed the backup of my 650GB fotos+videos from our travels in decent time 🎉

**Comment by gfelbing** [comment]
Hey @ncw ,

while my initial PR was open for a while, I gave the topic some more thoughts based on your review and my observations during when working with my fork:

As you already mentioned the pcloud API is pretty generic and instead of chunk writer we can simply implement OpenWriterAt, sparing a lot of duplicated logic. Nice!

Additionally, pcloud does show partially uploaded files (and leaves them broken on unfinished uploads), also in the current implementation.
The current implementation [tries to fix it](https://github.com/rclone/rclone/blob/27267547b903fe5e72aaa71cc1f45456661d0422/backend/pcloud/pcloud.go#L1238) by trying to delete on error, but this doesn't always work, e.g. when killing the process ungracefully or when the error is due to persistent connection loss. 

So IMHO setting the feature `PartialUploads` to true would be the cleanest option.
This brings one huge drawback to the chunked upload: We have no way to skip already uploaded chunks on partial files anymore, as the retry gets a new [random suffix](https://github.com/rclone/rclone/blob/27267547b903fe5e72aaa71cc1f45456661d0422/fs/operations/copy.go#L100).

# proposed solution

In order to fix this, I came up with a solution I'd like to discuss (will create a PR outlining it shortly):

1. Implement stable partial suffixes instead of random ones, by hashing the source info. 
2. Add a feature flag for backends that support resuming partial uploads, skipping deletion on error
3. Implement OpenWriterAt instead of OpenChunkedWriter for pcloud backend

With 1 and 2, we pcloud can cleanly set PartialUploads to true while keeping the resume functionality.
This could be useful for other backends as well, that also support resuming uploads but have PartialUploads enabled.

What do you think?

**Comment by gfelbing** [comment]
@ncw , as I removed the resume part from my PR, I'd like to move the discussion about the resume here.

Your comment on the PR:

> However I don't like the resume changes. Rclone backends don't resume, and it is very difficult to do and ensure data integrity (how are we 100% sure that the partial file uploaded is the same as the start of the file we are about to upload?).

This could be the responsibility of the backend, couldn't it? 
E.g. pcloud offers an API to get the checksum on any part of the file. When uploading a chunk, we can query the checksum first and skip if it fits.
Additionally, as a last line of defense, the copy operation still checks the integrity of the whole file after the resumed upload has finished.

Looking forward to your thoughts on this

**Comment by ncw** [comment]
I suggest you check out #87 and #4547 for a framework for doing this for lots of backends. This never got merged though so needs a champion to take it over.
